### PR TITLE
Make helm charts consistent with how fields in `spec` are handled. (kibana only)

### PR DESCRIPTION
**Pull request by naemono**
related: #8184

# Details of suggested change

It was pointed out in the above PR that the helm charts aren't consistent in how they implement values that end up within the `spec` field of the associated resource. Unfortunately the changes in 8184 are not backwards compatible, and are a breaking change. I went down the path of determining what a non-breaking change would look like, and just implemented this in the `eck-kibana` chart.

The question I have:
* Is this a path we want to go down to make the charts more consistent? Because this makes the templates more complex, but on the plus side, it significantly increases the test coverage.

(The intent is to do this for all inconsistent charts; just wanted to make the initial review/discussions easier)

### Implementation Note

1. The way this is implemented makes the previous behavior (having `.spec.field` in the values take precedence over just `field` in the values.
2. Mixing these 2 forms of values (some in spec in the values file, some not in spec) likely has unpredictable behavior. (for instance setting `spec.http.something`, and also setting `http.somethingElse`)

### CODE REVIEW

**Review by kvalliyurnatt — APPROVED** [review]

File: deploy/eck-stack/charts/eck-kibana/examples/http-configuration.yaml [hunk 1]
```diff
@@ -13,25 +13,24 @@ labels: {}
 annotations: {}
   # key: value
 
-spec:
```
Comment (kvalliyurnatt) [inline deploy/eck-stack/charts/eck-kibana/examples/http-configuration.yaml:16]: should we add a comment saying this spec. is deprecated and config. is the preferred way to set these configs. Not sure what the right place is to add this comment

Comment (kvalliyurnatt) [reply deploy/eck-stack/charts/eck-kibana/examples/http-configuration.yaml:16]: Thanks, LGTM.

**Comment by naemono** [comment]
> We probably also want to refactor [deploy/eck-stack/examples /kibana/](https://github.com/elastic/cloud-on-k8s/tree/main/deploy/eck-stack/examples/kibana.).

great catch! will.do

**Comment by naemono** [comment]
> > We probably also want to refactor [deploy/eck-stack/examples /kibana/](https://github.com/elastic/cloud-on-k8s/tree/main/deploy/eck-stack/examples/kibana.).
> 
> great catch! will.do

@thbkrkr this has been fixed.

File: deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml [hunk 2]
```diff
@@ -12,4 +12,49 @@ metadata:
...
   version: {{ required "A Kibana version is required" .Values.version }}
-  {{- toYaml .Values.spec | nindent 2 }}
+  {{- /* 
+    The following templates with 'or' are to allow both .spec.field and .field to be set for backwards 
+    compatibility purposes. See https://github.com/elastic/cloud-on-k8s/pull/8192 for details.
+  */ -}}
+  {{- with or ((.Values.spec).image) (.Values.image) }}
+  image: {{ . }}
+  {{- end }}
+  {{- with .Values.count }}
```
Comment (thbkrkr) [inline deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml:22]: I think we also want this field to be backwards compatible.
```suggestion
  {{- with or ((.Values.spec).count) (.Values.count) }}
```

File: deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml [hunk 3]
```diff
@@ -12,4 +12,49 @@ metadata:
...
+  {{- end }}
+  {{- with or ((.Values.spec).secureSettings) (.Values.secureSettings) }}
+  secureSettings:
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+  {{- with or ((.Values.spec).serviceAccountName) (.Values.serviceAccountName) }}
+  serviceAccountName: {{ . }}
+  {{- end }}
+  {{- with or ((.Values.spec).monitoring) (.Values.monitoring) }}
+  monitoring: {{ . }}
```
Comment (thbkrkr) [inline deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml:59]:
```suggestion
  monitoring:
    {{- toYaml . | nindent 4 }}
```

Comment (naemono) [reply deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml:59]: Let me actually get a test in place, as this should have been caught...

**Review by thbkrkr — APPROVED** [review]
LGTM

File: deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml (hunk 3 above)
Comment (naemono) [reply deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml]: Ok, new test showed the issue.
```
        - values.spec.monitoring should render properly

                - asserts[1] `equal` fail
                        Template:       eck-kibana/templates/kibana.yaml
                        DocumentIndex:  0
                        Path:   spec.monitoring
                        Expected to equal:
                                logs:
                                  elasticsearchRefs:
                                  - name: monitoring
                                    namespace: observability
                                metrics:
                                  elasticsearchRefs:
                                  - name: monitoring
                                    namespace: observability
                        Actual:
                                map[logs:map[elasticsearchRefs:[map[name:monitoring namespace:observability]]] metrics:map[elasticsearchRefs:[map[name:monitoring
                                  namespace:observability]]]]
```

And then fixed it

```
❯ helm unittest -3 -f 'templates/tests/*.yaml' 

### Chart [ eck-kibana ] .

 PASS  test kibana ingress      templates/tests/ingress_test.yaml
 PASS  test kibana      templates/tests/kibana_test.yaml

Charts:      1 passed, 1 total
Test Suites: 2 passed, 2 total
Tests:       21 passed, 21 total
Snapshot:    0 passed, 0 total
Time:        48.881008ms
```

File: deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml [hunk 4]
```diff
@@ -19,9 +19,7 @@ spec:
   {{- with or ((.Values.spec).image) (.Values.image) }}
   image: {{ . }}
   {{- end }}
-  {{- with or ((.Values.spec).count) (.Values.count) }}
-  count: {{ . }}
-  {{- end }}
+  count: {{ or ((.Values.spec).count) (.Values.count) | int | default 1 }}
```
Comment (thbkrkr) [inline deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml:22]: What is the reasoning for adding a default value in the template? Is the default value defined in the default values.yaml not enough?

Comment (naemono) [reply deploy/eck-stack/charts/eck-kibana/templates/kibana.yaml]: Ok, I was being overly cautious. I reverted this.

**Comment by naemono** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|2.16|[](https://github.com/elastic/cloud-on-k8s/pull/8289)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**